File: switchyard/lib/profiles/subagent_override.py
# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0

"""Route delegated sub-agent requests to a fixed worker runtime."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from switchyard.lib.request_metadata import CTX_PROFILE_REQUEST_HEADERS
from switchyard_rust.core import is_subagent_request

if TYPE_CHECKING:
    from switchyard.lib.proxy_context import ProxyContext
    from switchyard.lib.route_table import ChainRuntime
    from switchyard_rust.core import ChatRequest


class SubagentOverrideRuntime:
    """Wrap a runtime, sending delegated sub-agent work to ``worker`` instead.

    Normal traffic runs the wrapped runtime unchanged. A request whose headers carry a
    delegated sub-agent signal runs ``worker`` — a passthrough to the route's configured
    ``subagent_target``. The override never rewrites the request or response, and a worker
    failure surfaces as a normal target error rather than falling back to the wrapped
    runtime.

    Detection is :func:`~switchyard_rust.profiles.is_subagent_request`, the protocol
    crate's canonical policy, so harness maintenance turns (Codex ``compact``) stay on
    normal routing and every engine agrees on what counts as delegated work.
    """

    def __init__(self, inner: ChainRuntime, worker: ChainRuntime) -> None:
        self._inner = inner
        self._worker = worker

    def _branch(self, ctx: ProxyContext | None) -> ChainRuntime:
        """Select the worker for delegated work, else the wrapped runtime.

        Without a context there are no retained headers to read, so the request cannot be
        identified as delegated and routes normally.
        """
        if ctx is None:
            return self._inner
        headers = ctx.metadata.get(CTX_PROFILE_REQUEST_HEADERS) or {}
        return self._worker if is_subagent_request(headers) else self._inner

    async def call(self, request: ChatRequest, *, ctx: ProxyContext | None = None) -> Any:
        """Execute one request through the branch its headers select."""
        return await self._branch(ctx).call(request, ctx=ctx)

    def iter_components(self) -> list[Any]:
        """Return both branches' lifecycle components in startup order."""
        return [*self._inner.iter_components(), *self._worker.iter_components()]


__all__ = ["SubagentOverrideRuntime"]

File: crates/switchyard-components-v2/src/profiles/subagent_override.rs
// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

//! Sub-agent override combinator for v2 profiles.
//!
//! Wraps any profile without changing its behavior for normal traffic. A
//! request whose headers mark delegated sub-agent work
//! ([`switchyard_protocol::Metadata::is_subagent_work`]) is served directly by
//! a fixed worker target — keeping a sub-agent loop on an intentional,
//! cache-compatible target — while every other request delegates to the wrapped
//! profile. A worker failure surfaces as a normal target error and is never
//! silently re-routed through the wrapped profile.

use std::collections::BTreeMap;
use std::time::Instant;

use async_trait::async_trait;
use switchyard_components::StatsAccumulator;
use switchyard_core::{LlmTarget, Result};

use crate::backend::{native_target_backend, TargetBackend};
use crate::profile_stats_accumulator;
use crate::stats_recording::record_usage_or_wrap_stream;
use crate::{Profile, ProfileInput, ProfileResponse};

/// Wraps a profile, routing delegated sub-agent work to a fixed worker target.
pub(crate) struct SubagentOverrideProfile {
    inner: Box<dyn Profile>,
    worker: TargetBackend,
    stats: StatsAccumulator,
}

impl SubagentOverrideProfile {
    /// Wraps `inner`, routing recognized sub-agent work requests to `worker`.
    pub(crate) fn new(inner: Box<dyn Profile>, target: LlmTarget) -> Result<Self> {
        Ok(Self {
            inner,
            worker: native_target_backend(target)?,
            stats: profile_stats_accumulator(),
        })
    }
}

/// Returns true when the request headers signal delegated sub-agent work.
///
/// Normalizes multi-value headers to single-value (first occurrence wins)
/// before calling the protocol-layer detection logic.
fn is_subagent_work(headers: &BTreeMap<String, Vec<String>>) -> bool {
    let flat: BTreeMap<String, String> = headers
        .iter()
        .filter_map(|(k, vs)| vs.first().map(|v| (k.clone(), v.clone())))
        .collect();
    switchyard_protocol::Metadata::from_headers(&flat).is_subagent_work()
}

#[async_trait]
impl Profile for SubagentOverrideProfile {
    /// Routes sub-agent work to the worker target; delegates everything else.
    async fn run(&self, mut input: ProfileInput) -> Result<ProfileResponse> {
        if !is_subagent_work(&input.metadata.headers) {
            return self.inner.run(input).await;
        }

        let profile_started_at = Instant::now();
        let target_model = self.worker.target().model.clone();
        input.request.set_model(target_model.as_str());

        let backend_started_at = Instant::now();
        let response = match self.worker.call(&input.request).await {
            Ok(response) => response,
            Err(error) => {
                self.stats.record_error(target_model.as_str(), None)?;
                return Err(error);
            }
        };

        let backend_latency_ms = backend_started_at.elapsed().as_secs_f64() * 1000.0;
        self.stats
            .record_success(target_model.to_string(), Some(backend_latency_ms), None)?;
        let response = record_usage_or_wrap_stream(
            &self.stats,
            target_model.as_str(),
            None,
            profile_started_at,
            backend_latency_ms,
            response,
        )?;
        Ok(ProfileResponse::from(response))
    }
}
